### PR TITLE
feat(EMS-2612-2371): No PDF - Your Buyer - Total outstanding additional fields

### PR DESCRIPTION
**Pull request by Zainzzkk**
## Introduction :pencil2:
This PR adds the additional fields to the trading history page and adds validation for them

## Resolution :heavy_check_mark:
* Added additional fields via a conditional reveal
* Added validation for fields
* Added templates for conditional reveal including partials
* Added tests and e2e test

## Miscellaneous :heavy_plus_sign:
* Temporarily deleted totalOutstanding and amountOverdue from payload in map-submitted-data
* Split up trading-history tests
* Fixed failing date unit tests
* changed turnover to use shared partial for `provideAlternativeCurrencyLink`

### CODE REVIEW

File: e2e-tests/commands/insurance/your-buyer/assert-trading-history-outstanding-payments-currency.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * if no isoCode or alternativeCurrency selected
+     * checks GBP displayed
+     */
+  } else {
+    cy.checkText(field(TOTAL_OUTSTANDING).prefix(), currencySymbol);
+    cy.checkText(field(AMOUNT_OVERDUE).prefix(), currencySymbol);
+  }
+};
+
```
Comment (ttbarnes) [inline e2e-tests/commands/insurance/your-buyer/assert-trading-history-outstanding-payments-currency.js:72]: Finding this hard to follow, can we extract all form input and submission things into its own command like in other places? e.g `completeAndSubmitXForm` so this is purely asserting things?

File: e2e-tests/constants/field-ids/insurance/your-buyer/index.js [hunk 2]
```diff
@@ -15,6 +15,8 @@ export const YOUR_BUYER = {
   CONNECTION_WITH_BUYER_DESCRIPTION: 'connectionWithBuyerDescription',
   TRADED_WITH_BUYER: 'exporterHasTradedWithBuyer',
   OUTSTANDING_PAYMENTS: 'outstandingPayments',
+  TOTAL_OUTSTANDING: 'totalOutstanding',
+  AMOUNT_OVERDUE: 'amountOverdue',
```
Comment (ttbarnes) [inline e2e-tests/constants/field-ids/insurance/your-buyer/index.js:19]: What is the "outstanding total" and "amount "for"? In terms of data quality/clarity I think this needs to be more specific, maybe 

```
totalOutstandingPayments
overduePaymentsAmount /  totalOverduePayments
```

What do you think?

File: e2e-tests/commands/insurance/your-buyer/assert-trading-history-outstanding-payments-currency.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    cy.clickAlternativeCurrencyRadioOption();
+    cy.autocompleteKeyboardInput(ALTERNATIVE_CURRENCY_CODE, NON_STANDARD_CURRENCY_CODE);
+    cy.clickSubmitButton();
+
+    // clicks yes radio for total outstanding
+    yesRadioInput().first().click();
+
+    // asserts no prefix displayed
+    field(TOTAL_OUTSTANDING).prefix().should('not.exist');
+    field(AMOUNT_OVERDUE).prefix().should('not.exist');
```
Comment (ttbarnes) [inline e2e-tests/commands/insurance/your-buyer/assert-trading-history-outstanding-payments-currency.js:61]: Maybe we can create a generic command, `assertPrefixDoesNotExist`

As we're doing the same in some single and multiple contract policy tests (total contract value, export value)

File: e2e-tests/insurance/cypress/e2e/journeys/your-buyer/trading-history/trading-history-page.spec.js [hunk 4]
```diff
@@ -94,6 +93,60 @@ context('Insurance - Your Buyer - Trading history page - As an exporter, I want
...
+        it(`should not render an input for ${TOTAL_OUTSTANDING}`, () => {
+          field(TOTAL_OUTSTANDING).input().should('not.be.visible');
+        });
+
+        it(`should not render a label for ${AMOUNT_OVERDUE}`, () => {
+          field(AMOUNT_OVERDUE).label().should('not.be.visible');
+        });
+
+        it(`should not render an input for ${AMOUNT_OVERDUE}`, () => {
+          field(AMOUNT_OVERDUE).input().should('not.be.visible');
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/your-buyer/trading-history/trading-history-page.spec.js:115]: Should all of these tests be `not.exist`? I think visible would still pass if the element exists in the DOM.

File: src/ui/server/constants/templates/index.ts [hunk 5]
```diff
@@ -17,6 +17,7 @@ export const TEMPLATES = {
   ATTRIBUTES: {
     CLASSES: {
       LEGEND: {
+        S_NOT_BOLD: 'govuk-fieldset__legend--s govuk-!-font-weight-regular',
```
Comment (ttbarnes) [inline src/ui/server/constants/templates/index.ts:20]: Can we extract the "bold" as it's not specific to a legend

File: src/ui/server/controllers/insurance/your-buyer/map-submitted-data/buyer-trading-history/index.ts [hunk 6]
```diff
@@ -28,6 +31,10 @@ const mapSubmittedData = (formBody: RequestBody): object => {
     delete populatedData[ALTERNATIVE_CURRENCY_CODE];
   }
 
+  // TODO: Remove when data saving added for trading history
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/map-submitted-data/buyer-trading-history/index.ts:34]: ticket number?

File: src/ui/server/controllers/insurance/your-buyer/trading-history/validation/rules/amount-overdue.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+    // check if the field is below the minimum
+    if (Number(numberWithoutCommas) < MINIMUM) {
+      return generateValidationErrors(FIELD_ID, ERROR_MESSAGE.BELOW_MINIMUM, errors);
+    }
+  }
+
+  return updatedErrors;
+};
+
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/trading-history/validation/rules/amount-overdue.ts:50]: This means we have this logic duplicated in 3x places 😢 If you search for `wholeNumberValidation(formBody, updatedErrors, ERROR_MESSAGE.INCORRECT_FORMAT, FIELD_ID)`

Should we create a new DRY validation rule, maybe `wholeNumberAboveMinimum`? ANd just pass the body, field ID and an error message object to this

File: e2e-tests/constants/field-ids/insurance/your-buyer/index.js (hunk 2 above)
Comment (Zainzzkk) [reply e2e-tests/constants/field-ids/insurance/your-buyer/index.js:19]: done

File: e2e-tests/insurance/cypress/e2e/journeys/your-buyer/trading-history/trading-history-page.spec.js (hunk 4 above)
Comment (Zainzzkk) [reply e2e-tests/insurance/cypress/e2e/journeys/your-buyer/trading-history/trading-history-page.spec.js:115]: the other way around :P `not.exist` fails as it exists in the dom but `not.be.visible` works

Comment (ttbarnes) [reply e2e-tests/insurance/cypress/e2e/journeys/your-buyer/trading-history/trading-history-page.spec.js:115]: Ah gotcha, sorry I was thinking of the design scenario meaning it shouldn't exist on the page but this is for conditional reveals so make sense 👍

File: src/ui/server/controllers/insurance/your-buyer/trading-history/validation/rules/amount-overdue.ts (hunk 7 above)
Comment (Zainzzkk) [reply src/ui/server/controllers/insurance/your-buyer/trading-history/validation/rules/amount-overdue.ts:49]: done

File: e2e-tests/commands/insurance/complete-alternative-currency-form.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * completeAlternativeCurrencyForm
+ * presses hyperlink and completes the alternative currency form
+ * if isoCode provided - clicks the radio for currency and submits alternative currency form
+ * if alternativeCurrency true, then enters alternative currency and submits alternative currency form
+ * @param {String} isoCode - isoCode provided for radio selection
+ * @param {Boolean} alternativeCurrency - if alternative currency should be entered
+ */
+const completeAlternativeCurrencyForm = ({ isoCode, alternativeCurrency, linkFieldId }) => {
```
Comment (ttbarnes) [inline e2e-tests/commands/insurance/complete-alternative-currency-form.js:16]: `completeAndSubmit` ? It has submit button clicks

Maybe not for now, and not sure if this would be useful for this form, but in others we have `completeXForm` and `completeAndSubmit` so we can complete the form without submitting

File: e2e-tests/commands/shared-commands/assertions/assert-prefix.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import { field } from '../../../pages/shared';
+
+/**
+ * assertPrefix
+ * asserts prefix value or if it does not exist
+ * @param {String} fieldId fieldId for prefix
+ * @param {String} value prefix value
+ */
```
Comment (ttbarnes) [inline e2e-tests/commands/shared-commands/assertions/assert-prefix.js:8]: missing colons
`{String} fieldId: ...`
`{String} value: ...`

File: e2e-tests/commands/insurance/complete-alternative-currency-form.js (hunk 8 above)
Comment (Zainzzkk) [reply e2e-tests/commands/insurance/complete-alternative-currency-form.js:16]: don't think it'll be useful in this form

File: e2e-tests/insurance/cypress/e2e/journeys/your-buyer/trading-history/trading-history-change-currency.spec.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+const {
+  ROOT,
+  YOUR_BUYER: { TRADING_HISTORY },
+} = INSURANCE_ROUTES;
+
+const { AMOUNT_OVERDUE, TOTAL_OUTSTANDING } = FIELD_IDS;
+
+const baseUrl = Cypress.config('baseUrl');
+
+const linkFieldId = 'provide-alternative-currency';
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/your-buyer/trading-history/trading-history-change-currency.spec.js:17]: Can we use an actual `FIELD_ID`?

File: e2e-tests/commands/insurance/complete-alternative-currency-form.js (hunk 8 above)
Comment (ttbarnes) [reply e2e-tests/commands/insurance/complete-alternative-currency-form.js:16]: Ok can we rename to `completeAndSubmit ...` though because it reads like it's not submitting it, all other commands have "andSubmit" if it's submitting

File: src/ui/server/controllers/insurance/your-buyer/trading-history/validation/rules/amount-overdue.ts (hunk 7 above)
Comment (ttbarnes) [reply src/ui/server/controllers/insurance/your-buyer/trading-history/validation/rules/amount-overdue.ts:49]: Sorry if any confusion, there's still a lot of repitition with this validation logic. If I search for this:

```js
wholeNumberValidation(formBody, updatedErrors, ERROR_MESSAGE.INCORRECT_FORMAT, FIELD_ID);
```

There are now 5 validation functions with identital logic.

I think all of this repeated code:
```js
if (!objectHasProperty(formBody, FIELD_ID)) {
  return generateValidationErrors(FIELD_ID, ERROR_MESSAGE.INCORRECT_FORMAT, errors);
}

updatedErrors = wholeNumberValidation(formBody, updatedErrors, ERROR_MESSAGE.INCORRECT_FORMAT, FIELD_ID);

return wholeNumberAboveMinimumValidation(formBody, FIELD_ID, ERROR_MESSAGE.BELOW_MINIMUM, errors, MINIMUM);

if (belowMinimum) {
  return belowMinimum;
}
```

Could simply be in it's own shared function with reduces lines of code:

```js
if (!objectHasProperty(formBody, FIELD_ID)) {
  return generateValidationErrors(FIELD_ID, ERROR_MESSAGE.INCORRECT_FORMAT, errors);
}

return wholeNumberAboveMinimumValidation(formBody, FIELD_ID, ERROR_MESSAGES_OBJECT, errors, MINIMUM);

```

And then for each of the 5 files,  they can just call this new helper function with a field ID, `ERROR_MESSAGES_OBJECT` and maximum value in 1 line of code.

**Review by ttbarnes — APPROVED** [review]

**Review by ttbarnes — APPROVED** [review]
